### PR TITLE
Send movement changes

### PR DESCRIPTION
**Pull request by manucamejo**
## Motivation

We need to send to the backend only changes instead of 1 message every 30ms

## Summary of changes

- Improve client prediction

## How has this been tested?

Test it on Brazil and Europe. Check if the movement feels fluent or not.

## Checklist
- [x] I have tested the changes locally.
- [x] I self-reviewed the changes on GitHub, line by line.
- [ ] Tests have been added/updated.
- [x] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [x] I have tested the changes in another devices.
  - [x] Tested in iOS.
  - [ ] Tested in Android.

### CODE REVIEW

File: client/Assets/Scripts/Battle.cs [hunk 1]
```diff
@@ -261,9 +261,10 @@ void UpdatePlayerActions()
                     GameServerConnectionManager
                         .Instance
                         .clientPrediction
-                        .simulatePlayerState(
+                        .SimulatePlayerState(
                             serverPlayerUpdate,
-                            gameEvent.PlayerTimestamps[player.Id]
+                            gameEvent.PlayerTimestamps[player.Id],
```
Comment (BertovDev) [inline client/Assets/Scripts/Battle.cs:266]: We are keep sending `PlayerTimestamps` even when the player is dead.

Comment (BertovDev) [reply client/Assets/Scripts/Battle.cs:266]: We should add a conditional  `&& serverPlayerUpdate.Player.Health > 0` in the line `252`.

**Review by BertovDev — CHANGES_REQUESTED** [review]
This is working awesome and just a small comment.

**Review by BertovDev — COMMENTED** [review]
Tested on Android S23. Server Brazil and Europe.

**Review by BertovDev — DISMISSED** [review]

File: client/Assets/Scripts/ClientPrediction.cs [hunk 2]
```diff
@@ -8,71 +8,108 @@ public struct PlayerInput
...
+        UpdateLastAcknowledgedInput(player, timestampId, serverTimestamp);
+        RemoveServerAcknowledgedInputs(timestampId);
+        SimulatePlayerMovement(player);
+    }
+
+    void UpdateLastAcknowledgedInput(Entity player, long timestampId, long serverTimestamp)
     {
-        removeServerAcknowledgedInputs(player, timestamp);
-        simulatePlayerMovement(player);
+        for (int i = 0; i < pendingPlayerInputs.Count; i++)
```
Comment (agustinesco) [inline client/Assets/Scripts/ClientPrediction.cs:43]: I you're just using the index to get  the pending input you could use a foreach here

File: client/Assets/Scripts/Battle.cs [hunk 3]
```diff
@@ -249,7 +249,7 @@ void UpdatePlayerActions()
                 Entity serverPlayerUpdate = new Entity(gameEvent.Players[player.Id]);
                 if (
                     serverPlayerUpdate.Id == (ulong)GameServerConnectionManager.Instance.playerId
-                    && useClientPrediction
+                    && useClientPrediction && serverPlayerUpdate.Player.Health > 0
```
Comment (agustinesco) [inline client/Assets/Scripts/Battle.cs:252]: I think you can also check the player status here

**Review by agustinesco — APPROVED** [review]
LGTM! 
To tests this i used toxi-proxy to give myself some lag and it's working amazing! 
0ms observations: there's no difference between the main branch, the movement is smooth and accurate
150ms observations: The movement is smooth but at the end of the movement you can see the player move a bit back without animation
300ms observations: The movement is smooth but at the end of the movement you can see the player walks a bit back
1500ms observations: The client movement seems smooth but at the end of the movement you can see the player walks in the same place for a while

**Review by BertovDev — APPROVED** [review]
LGTM!!

**Pull request merged**